### PR TITLE
Unresolver: strict validation for external versions and other fixes

### PR DESCRIPTION
**Pull request by stsewd**
- At first I was thinking to add this validation
  outside the unresolver, but I think it makes
  sense to be in the unresolver.
- This also has a fix for partial subproject matches.
- Instead of returning the query and the fragment
  separately, just return the parsed URL, I think
  it can useful to have the full path without
  having to call urllib.parse again.
- Just return the extracted external version
  from unresolve_domain, so we don't have to parse the
  subdomain again.

This is on top of https://github.com/readthedocs/readthedocs.org/pull/9500.

---
:books: Documentation previews :books:

- User's documentation (`docs`): https://docs--9534.org.readthedocs.build/en/9534/

- Developer's documentation (`dev`): https://dev--9534.org.readthedocs.build/en/9534/

### CODE REVIEW

File: readthedocs/core/unresolver.py [hunk 1]
```diff
@@ -138,6 +158,9 @@ def _match_subproject(self, parent_project, path):
 
         If the subproject exists, we try to resolve the rest of the path
         with the subproject as the canonical project.
+
+        If the subproject exists, we return a result even if version doesn't,
+        so the subproject is taken as the current project (useful for 404 pages).
```
Comment (ericholscher) [inline readthedocs/core/unresolver.py:163]: Great comment 👍

**Review by ericholscher — COMMENTED** [review]
This looks like a good change, but I do worry a bit about some of the response magic that I don't fully understand here, and what it's handling.

File: readthedocs/core/unresolver.py [hunk 2]
```diff
@@ -82,6 +83,26 @@ def unresolve(self, url, add_index=True):
...
+        # Make sure we are serving the external version from the subdomain.
+        if external_version_slug and version:
+            if external_version_slug != version.slug:
+                log.warning(
+                    "Invalid version for external domain.",
+                    domain=domain,
+                    version_slug=version.slug,
+                )
+                version = None
+                filename = None
```
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/core/unresolver.py:95]: Setting version to `None` will be interpreted as like the version wasn't found, but we could return None, or raise an exception, since those URLs will be generated by manually changing the URL.

File: readthedocs/core/unresolver.py [hunk 3]
```diff
@@ -151,11 +174,15 @@ def _match_subproject(self, parent_project, path):
             .first()
         )
         if project_relationship:
-            return self._unresolve_path(
-                parent_project=project_relationship.child,
+            subproject = project_relationship.child
+            response = self._unresolve_path(
+                parent_project=subproject,
```
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/core/unresolver.py:179]: We use the subproject as the new parent project to unresolve the rest of the path, but the final parent project will always be the one we got from unresolving the domain

https://github.com/readthedocs/readthedocs.org/blob/c30087f95a32fec708f365801f94e0faef4d3d46/readthedocs/core/unresolver.py#L71-L77

this is so we can recursively unresolve the path.

**Comment by ericholscher** [comment]
I think this is close once we fix up the conflicts 👍

File: readthedocs/core/unresolver.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/core/unresolver.py:95]: I thought about this as well, and I'm not sure what's my position here.

I'd like to raise exceptions for all these cases were we return `None` for not found things. However, I'm not sure it's possible for all the cases because IIRC there are other places were `project, None, None` is useful to continue moving forward.

If that's not the case and I'm confused here, I'm 👍🏼 on standardizing this and always raise a detailed exception for each case that we can clearly handle from outside.

File: readthedocs/core/unresolver.py [hunk 4]
```diff
@@ -82,6 +83,26 @@ def unresolve(self, url, add_index=True):
...
+                log.warning(
+                    "Invalid version for external domain.",
+                    domain=domain,
+                    version_slug=version.slug,
+                )
+                version = None
+                filename = None
+            elif not version.is_external:
+                log.warning(
+                    "Version is not external.",
```
Comment (humitos) [inline readthedocs/core/unresolver.py:98]:
```suggestion
                    "Attempt to serve a non external external version from RTD_EXTERNAL_VERSION_DOMAIN.",
```

File: readthedocs/core/unresolver.py [hunk 5]
```diff
@@ -82,6 +83,26 @@ def unresolve(self, url, add_index=True):
...
+                    domain=domain,
+                    version_slug=version.slug,
+                    version_type=version.type,
```
Comment (humitos) [inline readthedocs/core/unresolver.py:101]: It would be good to log the full URL here as well.

File: readthedocs/core/unresolver.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/core/unresolver.py:95]: Actually, I think the case that I'm referring to is the 404 pages (see https://github.com/readthedocs/readthedocs.org/pull/9534/files#diff-3f2b884fdd4752fe36587aee4530694e7a4ff10a8f4523a7b0bc25e9d2f11e11R162)

File: readthedocs/core/unresolver.py [hunk 6]
```diff
@@ -249,23 +279,23 @@ def unresolve_domain(self, domain):
             # Serve from the PUBLIC_DOMAIN, ensuring it looks like `foo.PUBLIC_DOMAIN`.
             if public_domain == root_domain:
                 project_slug = subdomain
-                log.debug("Public domain.", domain=domain)
-                return project_slug, None, False
+                log.info("Public domain.", domain=domain)
```
Comment (humitos) [inline readthedocs/core/unresolver.py:282]:
```suggestion
                log.debug("Public domain.", domain=domain)
```

I think we don't want to change these logs. Probably a merge conflict.

File: readthedocs/core/unresolver.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/core/unresolver.py:179]: I got confused by this also in the previous PR as well. https://github.com/readthedocs/readthedocs.org/pull/9500#discussion_r964635220

It definitely sounds weird. We should, at least, add a comment here explaining why we are doing this. We both have been bitten by this already 😄

File: readthedocs/core/unresolver.py [hunk 7]
```diff
@@ -151,11 +174,15 @@ def _match_subproject(self, parent_project, path):
...
-            return self._unresolve_path(
-                parent_project=project_relationship.child,
+            subproject = project_relationship.child
+            response = self._unresolve_path(
+                parent_project=subproject,
                 path=file,
                 check_subprojects=False,
             )
+            if response:
+                return response
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/core/unresolver.py:187]: I mentioned this in the other review that I've done as well. See https://github.com/readthedocs/readthedocs.org/pull/9500#discussion_r964646110

File: readthedocs/core/unresolver.py [hunk 8]
```diff
@@ -216,7 +246,7 @@ def _unresolve_path(self, parent_project, path, check_subprojects=True):
             if response:
                 return response
 
-        return None, None, None
+        return parent_project, None, None
```
Comment (humitos) [inline readthedocs/core/unresolver.py:249]: Why we are returning the `parent_project` now? What changed in the logic?

File: readthedocs/core/unresolver.py [hunk 9]
```diff
@@ -249,23 +279,23 @@ def unresolve_domain(self, domain):
...
+            return None, None, None
 
         # Serve PR builds on external_domain host.
         if external_domain == root_domain:
             try:
-                log.debug("External versions domain.", domain=domain)
-                project_slug, _ = subdomain.rsplit("--", maxsplit=1)
-                return project_slug, None, True
+                project_slug, version_slug = subdomain.rsplit("--", maxsplit=1)
+                log.info("External versions domain.", domain=domain)
```
Comment (humitos) [inline readthedocs/core/unresolver.py:294]:
```suggestion
                log.debug("External versions domain.", domain=domain)
```

File: readthedocs/core/unresolver.py (hunk 8 above)
Comment (stsewd) [reply readthedocs/core/unresolver.py:249]: If nothing matches, the current project is the parent project, this was being handled outside, but having this here makes it easy easier (no need to do `project_result or project`).

File: readthedocs/core/unresolver.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/core/unresolver.py:106]: yeah, basically this would be interpreted as a 404, like if the version wouldn't exist.

File: readthedocs/core/unresolver.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/core/unresolver.py:201]: Added a comment for this

**Comment by stsewd** [comment]
I have updated this PR with some of comments from the review from
https://github.com/readthedocs/readthedocs.org/pull/9425.

There are a couple I didn't address, since I'm -0.
But we can continue discussing if you feel strongly.
I'd merge this for now, since all other things are related to refactoring.

### Renames

- `UnresolvedURL.project` -> `UnresolvedURL.current_project`.

  > Probably makes sense to change the UnresolvedURL to use current_project as
  > well so we keep consistency all across our code.

  I think it's fine to refer to it as just project inside the dataclass
  (similar we have just version instead of `current_version`)


- `_match_multiversion_project` -> `_match_multiversion_path`, `_match_subproject` -> `_match_subproject_path`.

  > I think a better name for this is _match_multiversion_url or
  > _match_multiversion_path since it works over the url/path

  > Same here, I think it's better to call this _match_subproject_path

  `_match_single_version_project` -> `_match_single_version_path`?

  For me all those methods work over the project and the path, so I'm fine with any name,
  but slightly preference for the current one.

### Dataclass instead of a tuple

- _unresolve_path and friends

  > It would be good to make this a dataclass or similar. Returning these
  > structured tuples is always hard to parse mentally and keep that in mind
  > while working with this code and switching between files

- unresolve_domain

  > It would be good to make this a dataclass or similar here as well and all
  > these related methods.

I started the implementation using 3 dataclasses,
but it felt complex instead of returning the 3 elements
(if we are returning more than 3, than I think it makes sense to use a dataclass),
since we already have one dataclass for the final result, having more makes it confusing where they are used,
and all of these are private methods, so we won't use them outside the class.

### Others

> Also, I just noticed that we are calling root_project in the design doc, but
> in the implementation we are just calling it project.

We are calling it parent project (https://github.com/readthedocs/readthedocs.org/pull/9500#discussion_r948533609),
I think it was my mistake while updating the design doc to use root project instead of parent project
(https://github.com/readthedocs/readthedocs.org/pull/9425/commits/2658ec6fc17894fbac89787f3bfc9231d04401e4),
we were using parent project before I updated the document.
